### PR TITLE
Add support for Sparkfun XRP

### PR DESCRIPTION
**Pull request by ThadHouse**
Adds support for https://www.sparkfun.com/products/22230
Copied mostly from the sparkfun-pro-micro

### CODE REVIEW

**Comment by ThadHouse** [comment]
@jannic Updated both this and my other 2 PR's as well.

File: boards/sparkfun-xrp/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    Gpio16 { name: servo_1 },
+    Gpio17 { name: servo_2 },
+    Gpio18 { name: qwiic_sda },
+    Gpio19 { name: qwiic_scl },
+    Gpio20 { name: range_trigger },
+    Gpio21 { name: range_echo },
+    Gpio22 { name: user_button },
+    Gpio23 { name: wireless_power },
+    Gpio24 { name: wireless_data },
+    Gpio25 { name: wireless_cs }
```
Comment (ithinuel) [inline boards/sparkfun-xrp/src/lib.rs:45]: Is this related to the pico-w?
Does the carrier board always gets delivered with a pico-w presoldered?

**Review by ithinuel — CHANGES_REQUESTED** [review]
Thank you for your contribution, this is greatly appreciated!

Could you please provide an example demonstrating (some of) the differentiators of this board?

Eg, Something like reading the IMU & printing some defmt traces and controling servos in response.

File: boards/sparkfun-xrp/src/lib.rs (hunk 1 above)
Comment (ThadHouse) [reply boards/sparkfun-xrp/src/lib.rs:45]: Yes, those are related to it being a Pico W, and yes it is always a Pico W.

**Comment by ThadHouse** [comment]
Can the examples be done in a later PR? Doing examples for this specifically are kind of invasive and I don't think should block at least the merging of this.

File: boards/sparkfun-xrp/src/lib.rs (hunk 1 above)
Comment (jannic) [reply boards/sparkfun-xrp/src/lib.rs:45]: In that case, does it make sense to provide a BSP at this time?
We don't provide a BSP for the Pico W either, because the unique feature of that board, the wireless connectivity, is not yet usable with rp2040-hal.

Comment (ThadHouse) [reply boards/sparkfun-xrp/src/lib.rs:45]: The rest of it works though. You don’t need the wireless functionality to work properly.

**Comment by ThadHouse** [comment]
The main thing to figure out for the examples is many of them need special functionality. Like the quadrature encoders are a PIO program. That would make sense as a library, but not clear yet if it makes sense to go into the hal, have a new library with each individual thing (This smells way too JS like and I hate it), put it in the BSP, or have it just be a part of the examples.

**Comment by ithinuel** [comment]
I initiated a conversation on [`matrix`](https://matrix.to/#/!zbjndUiYJDvhDRkPGw:matrix.org/$DKNctTSdSf5qTlQktdVKp7-CqKOgpF2zLGTY_FMhI4Q?via=matrix.org&via=tchncs.de&via=boiler.social) to gather other people's thoughts.

I don't have strong opinions about the pins for wireless not being used yet because of lack of wifi support.

We are all volunteers with other day jobs. When we accept a crate (a BSP here, or any crate under rp-rs' org), we also (implicitly?) commit to maintaining it and keeping it current with the HAL's and other dependencies' updates.
So the issue I see is that accepting more and more BSPs increases the work load the community has to go through for updates/releases, dutsing it off from time to time.

With this in mind, I'd just like to make sure that when we accept to take care of/"own" a crate, it must at least bring something new/different.
Otherwise people may as-well use an existing crate with the same features but slightly different names.
This is only to "safe-guard" our time and ensure it brings a bit more value to the community.

Others may disagree, and I'm fine to have this merged _if they do_.

**Comment by ithinuel** [comment]
> […] Like the quadrature encoders are a PIO program. That would make sense as a library, 
but not clear yet if it makes sense to go into the hal, have a new library with each individual […]

This is indeed how PIO features are usually exposed (see [i2c-pio](https://github.com/rp-rs/i2c-pio-rs), [spi-pio](https://github.com/rp-rs/spi-pio-rs), [ws2812-pio](https://github.com/rp-rs/ws2812-pio-rs)).

> put it in the BSP, or have it just be a part of the examples.

If this is tiny and can fit in 10-20lines, that would be fine IMHO.

If quadrature encoder are too much work, there's still:

- https://crates.io/crates/servo-pio: Servo control using PIO
- https://crates.io/crates/lsm6dso: IMU drivers

I haven't looked at the details of those drivers but they don't need to be feature complete to be used in an example.
If at some point a user wants to update them to showcase a more feature complete driver that's fine too :)

**Comment by ThadHouse** [comment]
I think this is actually one BSP my team would probably want to have our own repo for. Theres a lot of functionality on this board and its probably better to have it be standalone.

**Comment by jannic** [comment]
We could still add a link to that bsp to the readme if you like.
